### ISSUE TITLE
Quartz cannot find bean instances of jobs

### ISSUE THREAD

**Issue by ferencbeutel4711** [quarkusio/quarkus#18885]
### Describe the bug

Hello there,

we have an issue that is present with quarkus 1.13 - 2.0.2, in 1.12 the issue is not yet present.

In our setup we are using quartz to schedule the execution of jobs. Here is a simple implementation example of one of our jobs:

```
@Startup
@Singleton
class HouseKeepingJob(
  scheduler: Scheduler,
  jobRepository: JobRepository,
  @ConfigProperty(name = "job.timeout") private val timeout: Long,
) : BaseJob(
  jobRepository = jobRepository,
  scheduler = scheduler,
) {

  override fun execute() {
    // implementation details
  }

  override fun getCron() {
    // implementation details
  }
}

class BaseJob(
  scheduler: Scheduler,
  jobRepository: JobRepository,
) {
  init {
    val job = JobBuilder.newJob(this::class.java).build()
    val trigger: Trigger = TriggerBuilder.newTrigger()
      .withSchedule(CronScheduleBuilder.cronSchedule(getCron()))
      .build()

    scheduler.scheduleJob(job, trigger)
  }

  abstract fun getCron(): String
}
```

As you can see in the above example, we are creating a Singleton instance of our job which should be used by quartz. With quarkus 1.12.x, this is working perfectly fine. From 1.13.x above, we have the issue that in the `InvokerJobFactory` deep inside quarkus-quartz, the singleton instance of our job cannot be resolved anymore, therefore a fallback is used (`SimpleJobFactory`) which fails due to a missing no-args constructor on our job.
We have found that the `jobClass` used for bean resolving differs between 1.12 and 1.13 - while in 1.12 the jobClass would be `HouseKeepingJob` in 1.13 it is `HouseKeepingJob_SubClass`, which is the only major difference we were able to spot.
In the migration guide we cannot find any info concerning such a change. Can you please help us understand what changed and what the best course of action is for us right now?

Thanks

### Expected behavior

the singleton instance of our job should still be discoverable in 1.13

### Actual behavior

the singleton instance of our job is not discoverable anymore since 1.13

### Output of `uname -a` or `ver`

Darwin 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64

### Output of `java -version`

openjdk version "11.0.9.1" 2020-11-04 LTS OpenJDK Runtime Environment Zulu11.43+1021-CA (build 11.0.9.1+1-LTS) OpenJDK 64-Bit Server VM Zulu11.43+1021-CA (build 11.0.9.1+1-LTS, mixed mode)

### Quarkus version or git rev

1.13.7

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

gradle 6.7.1

**Comment by geoand** [comment]
Can you add a reproducer project please?

**Comment by ferencbeutel4711** [comment]
hi geoand,

sure, no problem. You can find it here: https://github.com/ferencbeutel4711/repro-18885

just run it with quarkusDev, after roughly 15 seconds you will see the exception.

**Comment by geoand** [comment]
Thanks!

I don't know much about Quarkus, but I'll have a look anyway. If I can't figure it out, I'll pull in others who have actually worked in this area.

**Comment by geoand** [comment]
The easy workaround for now is to add: `quarkus.arc.dev-mode.monitoring-enabled=false` you `application.properties`. More information about this can be found [here](https://quarkus.io/guides/cdi-reference#quarkus-arc_quarkus.arc.dev-mode.monitoring-enabled).

**Comment by geoand** [comment]
It looks like Arc should not be trying to create a proxy in this case, but in any case that's one for @mkouba

**Comment by machi1990** [comment]
> I don't know much about Quarkus, but I'll have a look anyway.... 

@geoand sure you do :-)


Anyway, thanks for having a look at the issue.

**Comment by geoand** [comment]
> > I don't know much about Quarkus, but I'll have a look anyway....
> 
> @geoand sure you do :-)
> 
> Anyway, thanks for having a look at the issue.

Haha, I meant Quartz :P

**Comment by mkouba** [comment]
Ok, so the exception comes from the `org.quartz.simpl.SimpleJobFactory` which attempts to instantiate a `TestJob_Subclass` via a no-args constructor but the subclass obviously does not declare a no-args constructor. It probably works with  `quarkus.arc.dev-mode.monitoring-enabled=false` because the original class is transformed with the `allopen` plugin (because in fact the `HouseKeepingJob` class itself does not declare a no-args constructor).

> It looks like Arc should not be trying to create a proxy in this case (as there is no default constructor), but in any case that's one for @mkouba

ArC is not trying to create a proxy but an intercepted subclass which is needed in the dev mode, because by default all beans are monitored. In most cases, we do add a no-args constructor automatically but in this particular case:
1. It's not needed because intercepted subclasses work just fine with non-private `@Inject` constructors with params (a single constructor is considered an `@Inject` constructor)
2. The extended class (`BaseJob`) does not declare a no-args constructor either so we can't do this automatically.

To sum it up: ArC itself would work just fine but the way the job is instantiated relies on the `allopen` plugin functionality which we cannot reflect because we analyze the original classes. I'd recommend to try to switch to the [quarkus-quartz extension](https://quarkus.io/guides/quartz#scheduling-jobs-programmatically) which has a specialized `JobFactory` that should handle this situation correctly.

**Comment by geoand** [comment]
So should we close this @mkouba ?

**Comment by mkouba** [comment]
I'd like to close this issue if (1) @ferencbeutel4711 is happy with the workaround, or (2) migration to `quarkus-quartz` would solve the problem.

**Comment by ferencbeutel4711** [comment]
Hi @mkouba, thanks for your reply.

I am pretty sure we are using the `quarkus-quartz` dependency in the repro repo as seen here:

```
    implementation 'io.quarkus:quarkus-quartz'
```
https://github.com/ferencbeutel4711/repro-18885/blob/master/build.gradle#L14

Also, I dont understand what changed from 12.x -> 13.x that caused this issue to arise.

From my understanding, in the quarkus-quartz extension, there is the `InvokerJobFactory`, which is injected by the extension as the underlying jobFactory for quartz, as seen in line 109 of that class. The implementation of it looks good to me in general. 

In the repro example with quarkus 12.x, the `InvokerJobFactory` is using the jobName `HouseKeepingJob` to find an instance of the job, which is successful, since the singleton was already instantiated by quarkus and was injected as the `jobs` constructor argument of the `QuartzScheduler` from the quarkus extension.
In 13.x however, the jobName changed to `HouseKeepingJob_Subclass`, which is not found anymore in the `jobs` injected into the `QuartzScheduler` - however, when debugging, I've used `HouseKeepingJob` as the jobName, which successfully retrieved the job instance.

So from my POV, I am not sure if I have communicated the issue successfully in my posts earlier since the solutions do not really match to the issue, but maybe I just dont understand the answer well enough.

**Comment by geoand** [comment]
> Hi @mkouba, thanks for your reply.
> 
> I am pretty sure we are using the `quarkus-quartz` dependency in the repro repo as seen here:
> 
> ```
>     implementation 'io.quarkus:quarkus-quartz'
> ```
> 
> https://github.com/ferencbeutel4711/repro-18885/blob/master/build.gradle#L14
> 
> Also, I dont understand what changed from 12.x -> 13.x that caused this issue to arise.
> 

This is happening for the reason I explained above about the moniroting functionality Arc added.

> From my understanding, in the quarkus-quartz extension, there is the `InvokerJobFactory`, which is injected by the extension as the underlying jobFactory for quartz of the class `QuartzScheduler`. The implementation of it looks good to me in general.
> for reference: https://github.com/quarkusio/quarkus/blob/main/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L112
> 
> In the repro example with quarkus 12.x, the `InvokerJobFactory` is using the jobName `HouseKeepingJob` to find an instance of the job, which is successful, since the singleton was already instantiated by quarkus and was injected as the `jobs` constructor argument of the `QuartzScheduler` from the quarkus extension.
> In 13.x however, the jobName changed to `HouseKeepingJob_Subclass`, which is not found anymore in the `jobs` injected into the `QuartzScheduler` - however, when debugging, I've used `HouseKeepingJob` as the jobName, which successfully retrieved the job instance.
> 
> So from my POV, I am not sure if I have communicated the issue successfully in my posts earlier since the solutions do not really match to the issue, but maybe I just dont understand the answer well enough.

@mkouba the issue is that if monitoring is disabled, everything works as expected. If not, Arc is not resolving the job this leading to the failure of the default implementation

**Comment by mkouba** [comment]
> In 13.x however, the jobName changed to `HouseKeepingJob_Subclass`, which is not found anymore in the `jobs` injected into the `QuartzScheduler` - however, when debugging, I've used `HouseKeepingJob` as the jobName, which successfully retrieved the job instance.

I see. Well, the problem is that the job class is determined using the `this::class.java` which is really `HouseKeepingJob_Subclass` (and will be an subclass instance for any intercepted bean). One way to fix this problem is to detect an ArC subclass via ` io.quarkus.arc.Subclass.isAssignableFrom(jobClass)` and if so, then simply use the result of `java.lang.Class.getSuperclass()`.



> > Also, I dont understand what changed from 12.x -> 13.x that caused this issue to arise.
> 
> This is happening for the reason I explained above about the moniroting functionality Arc added.

Exactly, but keep in mind that you would get the same error if the bean that represents the job is intercepted by any interceptor (logging, transactions, etc.).

> > So from my POV, I am not sure if I have communicated the issue successfully in my posts earlier since the solutions do not really match to the issue, but maybe I just dont understand the answer well enough.
> 
> @mkouba the issue is that if monitoring is disabled, everything works as expected. If not, Arc is not resolving the job this leading to the failure of the default implementation

Yes, now fully understand the problem and I tried to describe the solution above ^^.  Note that `quarkus.arc.dev-mode.monitoring-enabled=false` will only work until another interceptor comes into play...

**Comment by geoand** [comment]
Should I assign this to you @mkouba ?

**Comment by mkouba** [comment]
I've just sent a PR - you can mimic [this logic](https://github.com/quarkusio/quarkus/pull/18931/files#diff-c639bfa620c120ae00ffdef3e0c3ef8dfc051f4f305c818a8691edb8e4c0bd31R493-R496) as a workaround.

**Comment by ferencbeutel4711** [comment]
hey guys, thank you so much for the quick fix and great work in general on this issue, it is much appreciated :)
